src/coreclr/jit/lower.cpp

```diff
@@ -5551,6 +5551,19 @@ GenTree* Lowering::LowerDelegateInvoke(GenTreeCall* call)
     BlockRange().Remove(thisArgNode);
     BlockRange().InsertBefore(call, newThisAddr, newThis, thisArgNode);
 
+#if !defined(TARGET_XARCH)
+    if (comp->GetInterruptible())
+    {
+        // If the target's backend doesn't support indirect calls with immediate operands (contained)
+        // and the method is marked as interruptible, we need to insert a GT_START_NONGC before the call.
+        // to keep the delegate object alive while we're obtaining the function pointer.
+        GenTree* startNonGCNode = new (comp, GT_START_NONGC) GenTree(GT_START_NONGC, TYP_VOID);
+        GenTree* stopNonGCNode  = new (comp, GT_STOP_NONGC) GenTree(GT_STOP_NONGC, TYP_VOID);
+        BlockRange().InsertAfter(thisArgNode, startNonGCNode);
+        BlockRange().InsertAfter(call, stopNonGCNode);
+    }
+#endif
```

Comment:
I would still do this inside `LowerCall` since it's `LowerCall` responsibility to insert the control expression returned by this function. This is adding an assumption about where `LowerCall` will insert that node.

Reply:
> I would still do this inside `LowerCall` since it's `LowerCall` responsibility to insert the control expression returned by this function. This is adding an assumption about where `LowerCall` will insert that node.

but that's the point, no? `LowerDelegateInvoke` already inserts some trees and at the end of those we conservatively emit a START_NOGC so other functions like `LowerCall` itself, `LowerCFGCall` or `LowerTailCallViaJitHelper` can insert whatever they want after our NOGC and before the call

Reply:
Hmm yeah, that's a good point too. What I dislike here is just that it still relies on `LowerCall` to insert the control expression _after_ the `START_NONGC`. To be conservatively correct we have to insert the `START_NOGC` _before_ the last use of the delegate, since `emitDisableGC()` means something like "the next instruction I emit is going to make GC information wrong, so disable GC after that instruction". Hence inserting `START_NONGC` after `thisArgNode` is only correct because there ends up being another use inserted after that point by `LowerCall`. The main idea behind moving the logic into `LowerCall` is that it has the full overview of what nodes it is going to insert and where, so it can make the decision with more information. So for instance it could insert it before the `gtControlExpr` in the delegate case, knowing that `gtControlExpr` has the last use of the delegate instance in the common case.